File: git_service.py
import os
from typing import List, Optional, Tuple
from urllib.parse import quote
from git import Repo, GitCommandError
from datetime import datetime
from utils.logger import Logger, LogLevel


class GitService:
    def __init__(self, config):
        self.config = config
        self.logger = Logger()
        self.repo: Optional[Repo] = None
        self.repo_path = os.path.join(os.path.expanduser("~"), ".auto-commit-repo")

    def _get_auth_url(self, url: str) -> str:
        """构建带认证的URL"""
        if not url.startswith(("http://", "https://")):
            return url

        # 移除已有的用户名密码部分
        if "@" in url:
            url = url.rsplit("@", 1)[1]

        # 统一转为 https 并提取 host+path
        if url.startswith("http://"):
            host_path = url[7:]
        elif url.startswith("https://"):
            host_path = url[8:]
        else:
            host_path = url

        encoded_user = quote(self.config.username, safe='')
        encoded_pass = quote(self.config.password, safe='')
        return f"https://{encoded_user}:{encoded_pass}@{host_path}"

    def clone_or_pull(self) -> bool:
        """克隆或更新仓库"""
        try:
            if not os.path.exists(self.repo_path):
                self.logger.info("正在克隆仓库...")
                auth_url = self._get_auth_url(self.config.repo_url)
                self.repo = Repo.clone_from(auth_url, self.repo_path)
                self.logger.success("仓库克隆成功")
                return True

            self.logger.info("正在更新仓库...")
            self.repo = Repo(self.repo_path)
            auth_url = self._get_auth_url(self.config.repo_url)

            # 配置远程URL
            if self.repo.remotes:
                origin = self.repo.remotes.origin
                origin.set_url(auth_url)

            self.repo.remotes.origin.fetch()
            self.logger.success("仓库更新成功")
            return True

        except GitCommandError as e:
            self.logger.error(f"Git操作失败: {e}")
            return False
        except Exception as e:
            self.logger.error(f"未知错误: {e}")
            return False

    def _resolve_remote_branch(self, branch_name: str) -> str:
        """检查远程分支是否存在，不存在则尝试检测默认分支"""
        remote_refs = [ref.name for ref in self.repo.references]
        target_ref = f"origin/{branch_name}"

        if target_ref in remote_refs:
            return branch_name

        # 尝试通过远程 HEAD 检测默认分支
        try:
            head_ref = self.repo.git.symbolic_ref('refs/remotes/origin/HEAD')
            # 返回类似 refs/remotes/origin/master
            default_branch = head_ref.replace('refs/remotes/origin/', '')
            self.logger.info(f"远程分支 '{branch_name}' 不存在，检测到默认分支: '{default_branch}'")
            return default_branch
        except GitCommandError:
            pass

        # 列出所有远程分支供参考
        remote_branches = [ref.name for ref in self.repo.references
                           if ref.name.startswith('origin/')]
        self.logger.error(
            f"远程分支 '{branch_name}' 不存在。"
            f"可用分支: {', '.join(remote_branches)}"
        )
        return branch_name

    def fetch_commits(self, since_hash: str = "") -> List[object]:
        """获取从指定hash开始的提交列表"""
        try:
            self.repo.git.fetch("--all")

            # 自动解析实际存在的远程分支
            actual_branch = self._resolve_remote_branch(self.config.source_branch)
            ref = f"origin/{actual_branch}"

            if not since_hash:
                # 如果是第一次同步，获取所有提交
                commits = list(self.repo.iter_commits(ref))
            else:
                # 获取从上次同步hash到最新的提交
                commits = []
                found_start = False
                for commit in self.repo.iter_commits(ref):
                    if commit.hexsha == since_hash:
                        found_start = True
                        break
                    commits.append(commit)

                # 如果没找到起始hash，获取所有提交
                if not found_start:
                    self.logger.warning("未找到上次同步的hash，将获取所有提交")
                    commits = list(self.repo.iter_commits(ref))

            # 反转顺序，从旧到新
            return list(reversed(commits))

        except GitCommandError as e:
            self.logger.error(f"获取提交记录失败: {e}")
            return []
        except Exception as e:
            self.logger.error(f"未知错误: {e}")
            return []

    def _target_branch_exists(self) -> bool:
        """检查目标分支是否已存在（本地或远程）"""
        local_branches = [b.name for b in self.repo.branches]
        if self.config.target_branch in local_branches:
            return True
        remote_refs = [ref.name for ref in self.repo.references]
        return f"origin/{self.config.target_branch}" in remote_refs

    def prepare_branch_full_sync(self) -> bool:
        """首次同步：基于源分支创建孤立的目标分支"""
        try:
            actual_branch = self._resolve_remote_branch(self.config.source_branch)
            self.repo.git.checkout(f'origin/{actual_branch}')

            # 删除已有的本地目标分支
            try:
                self.repo.git.branch('-D', self.config.target_branch)
            except GitCommandError:
                pass

            # 创建孤立分支（无历史记录）
            self.repo.git.checkout('--orphan', self.config.target_branch)
            self.repo.git.rm('-rf', '--cached', '.')
            self.repo.git.clean('-fd')

            self.logger.info(f"已创建孤立分支: {self.config.target_branch}")
            return True

        except GitCommandError as e:
            self.logger.error(f"创建分支失败: {e}")
            return False

    def prepare_branch_incremental(self) -> bool:
        """增量同步：切换到已存在的目标分支"""
        try:
            local_branches = [b.name for b in self.repo.branches]
            if self.config.target_branch in local_branches:
                self.repo.git.checkout(self.config.target_branch)
            else:
                # 本地不存在，从远程拉取
                self.repo.git.checkout(
                    f'origin/{self.config.target_branch}',
                    b=self.config.target_branch
                )

            self.logger.info(f"已切换到目标分支: {self.config.target_branch}")
            return True

        except GitCommandError as e:
            self.logger.error(f"切换分支失败: {e}")
            return False

    def _build_author_env(self, commit) -> dict:
        """构建修改作者的 Git 环境变量"""
        original_date = datetime.fromtimestamp(commit.committed_date).isoformat()
        return {
            'GIT_AUTHOR_DATE': original_date,
            'GIT_COMMITTER_DATE': original_date,
            'GIT_AUTHOR_NAME': self.config.user_name,
            'GIT_AUTHOR_EMAIL': self.config.user_email,
            'GIT_COMMITTER_NAME': self.config.user_name,
            'GIT_COMMITTER_EMAIL': self.config.user_email,
        }

    def recommit_full(self, commits: List[object]) -> Tuple[bool, int]:
        """全量同步：用 read-tree 回放所有提交（修改作者，保留时间）"""
        try:
            success_count = 0

            for commit in commits:
                try:
                    # 用该提交的树替换整个工作区和索引
                    self.repo.git.read_tree('-u', '--reset', commit.hexsha)

                    message = commit.message.strip() or f"Sync commit {commit.hexsha[:8]}"
                    env = self._build_author_env(commit)
                    self.repo.git.commit(m=message, env=env)
                    success_count += 1
                    self.logger.info(f"同步提交: {message[:50]}...")

                except GitCommandError as e:
                    self.logger.error(f"重新提交失败 {commit.hexsha[:8]}: {e}")
                    continue

            self.logger.success(f"成功同步 {success_count}/{len(commits)} 个提交")
            return True, success_count

        except Exception as e:
            self.logger.error(f"重新提交过程出错: {e}")
            return False, 0

    def recommit_incremental(self, commits: List[object]) -> Tuple[bool, int]:
        """增量同步：用 cherry-pick 回放新提交（修改作者，保留时间）"""
        try:
            success_count = 0

            for commit in commits:
                try:
                    message = commit.message.strip() or f"Sync commit {commit.hexsha[:8]}"
                    env = self._build_author_env(commit)

                    # cherry-pick 不提交，然后手动提交以覆盖作者信息
                    self.repo.git.cherry_pick('--no-commit', commit.hexsha)
                    self.repo.git.commit(m=message, env=env)
                    success_count += 1
                    self.logger.info(f"同步提交: {message[:50]}...")

                except GitCommandError as e:
                    self.logger.error(f"重新提交失败 {commit.hexsha[:8]}: {e}")
                    try:
                        self.repo.git.cherry_pick('--abort')
                    except GitCommandError:
                        pass
                    continue

            self.logger.success(f"成功同步 {success_count}/{len(commits)} 个提交")
            return True, success_count

        except Exception as e:
            self.logger.error(f"重新提交过程出错: {e}")
            return False, 0

    def push_to_remote(self) -> bool:
        """推送到远程仓库"""
        try:
            auth_url = self._get_auth_url(self.config.repo_url)
            self.repo.remotes.origin.set_url(auth_url)
            self.repo.git.push('origin', self.config.target_branch, '--force')
            self.logger.success("推送到远程成功")
            return True

        except GitCommandError as e:
            self.logger.error(f"推送失败: {e}")
            return False
        except Exception as e:
            self.logger.error(f"推送过程出错: {e}")
            return False

    def sync(self) -> bool:
        """执行完整同步流程"""
        self.logger.info("=" * 50)
        self.logger.info("开始同步...")

        # 克隆或更新仓库
        if not self.clone_or_pull():
            return False

        # 获取提交记录
        since_hash = self.config.last_sync_hash
        commits = self.fetch_commits(since_hash)

        if not commits:
            self.logger.info("没有新的提交需要同步")
            self.config.last_sync_time = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            return True

        self.logger.info(f"发现 {len(commits)} 个新提交")

        # 判断全量同步还是增量同步：目标分支必须同时存在于本地和远程才算增量
        is_full_sync = not since_hash or not self._target_branch_exists()

        if is_full_sync:
            if not self.prepare_branch_full_sync():
                return False
            success, count = self.recommit_full(commits)
        else:
            if not self.prepare_branch_incremental():
                return False
            success, count = self.recommit_incremental(commits)

        if not success:
            return False

        # 推送到远程
        if not self.push_to_remote():
            return False

        # 更新最后同步的hash
        last_commit = commits[-1]
        self.config.last_sync_hash = last_commit.hexsha
        self.config.last_sync_time = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        self.config.sync()

        self.logger.success(f"同步完成！共同步 {count} 个提交")
        self.logger.info("=" * 50)

        return True

File: src/services/git-service.js
const simpleGit = require('simple-git');
const path = require('path');
const os = require('os');
const fs = require('fs');

class GitService {
  constructor(config, logger) {
    this.config = config;
    this.logger = logger;
    this.sourceRepoPath = path.join(os.homedir(), '.auto-commit-source');
    this.targetRepoPath = path.join(os.homedir(), '.auto-commit-target');
    this.sourceGit = null;
    this.targetGit = null;
  }

  _getAuthUrl(url) {
    if (!url.startsWith('http://') && !url.startsWith('https://')) return url;

    let hostPath = url;
    if (hostPath.includes('@')) {
      hostPath = hostPath.split('@').pop();
    }
    if (hostPath.startsWith('http://')) {
      hostPath = hostPath.slice(7);
    } else if (hostPath.startsWith('https://')) {
      hostPath = hostPath.slice(8);
    }

    const encodedUser = encodeURIComponent(this.config.username);
    const encodedPass = encodeURIComponent(this.config.password);
    return `https://${encodedUser}:${encodedPass}@${hostPath}`;
  }

  async _cloneOrPull(repoUrl, localPath, label) {
    try {
      const authUrl = this._getAuthUrl(repoUrl);

      if (!fs.existsSync(localPath)) {
        this.logger.info(`正在克隆${label}仓库...`);
        await simpleGit().clone(authUrl, localPath);
        this.logger.success(`${label}仓库克隆成功`);
        return simpleGit(localPath);
      }

      this.logger.info(`正在更新${label}仓库...`);
      const git = simpleGit(localPath);
      await git.remote(['set-url', 'origin', authUrl]);
      await git.fetch('origin');
      this.logger.success(`${label}仓库更新成功`);
      return git;
    } catch (e) {
      this.logger.error(`${label}仓库操作失败: ${e.message}`);
      return null;
    }
  }

  async cloneOrPullSource() {
    this.sourceGit = await this._cloneOrPull(this.config.sourceRepoUrl, this.sourceRepoPath, '源');
    return this.sourceGit !== null;
  }

  async cloneOrPullTarget() {
    this.targetGit = await this._cloneOrPull(this.config.targetRepoUrl, this.targetRepoPath, '目标');
    return this.targetGit !== null;
  }

  async _resolveRemoteBranch(git, branchName) {
    try {
      const refs = await git.branch(['-r']);
      const remoteBranches = refs.all;
      const targetRef = `origin/${branchName}`;

      if (remoteBranches.includes(targetRef)) return branchName;

      try {
        const headRef = await git.raw(['symbolic-ref', 'refs/remotes/origin/HEAD']);
        const defaultBranch = headRef.trim().replace('refs/remotes/origin/', '');
        this.logger.info(`远程分支 '${branchName}' 不存在，检测到默认分支: '${defaultBranch}'`);
        return defaultBranch;
      } catch {}

      this.logger.error(`远程分支 '${branchName}' 不存在。可用分支: ${remoteBranches.join(', ')}`);
      return branchName;
    } catch {
      return branchName;
    }
  }

  async fetchCommits(sinceHash = '') {
    try {
      await this.sourceGit.fetch(['--all']);
      const actualBranch = await this._resolveRemoteBranch(this.sourceGit, this.config.sourceBranch);
      const ref = `origin/${actualBranch}`;

      const SEP = '__|__';
      const rawLog = await this.sourceGit.raw([
        'log', ref, `--format=%H${SEP}%at${SEP}%s`,
      ]);

      if (!rawLog || !rawLog.trim()) return [];

      const allCommits = rawLog.trim().split('\n').map(line => {
        const [hash, timestamp, ...msgParts] = line.split(SEP);
        return { hash, date: timestamp, message: msgParts.join(SEP) };
      });

      if (!sinceHash) {
        return allCommits.reverse();
      }

      const idx = allCommits.findIndex(c => c.hash === sinceHash);
      if (idx === -1) {
        this.logger.warning('未找到上次同步的hash，将获取所有提交');
        return allCommits.reverse();
      }

      return allCommits.slice(0, idx).reverse();
    } catch (e) {
      this.logger.error(`获取提交记录失败: ${e.message}`);
      return [];
    }
  }

  async _targetBranchExists() {
    try {
      const localBranches = await this.targetGit.branchLocal();
      if (localBranches.all.includes(this.config.targetBranch)) return true;

      const remoteBranches = await this.targetGit.branch(['-r']);
      return remoteBranches.all.includes(`origin/${this.config.targetBranch}`);
    } catch {
      return false;
    }
  }

  async _exportPatch(hash) {
    try {
      const patch = await this.sourceGit.raw(['diff-tree', '-p', hash]);
      return patch;
    } catch (e) {
      this.logger.error(`导出补丁失败 ${hash.slice(0, 8)}: ${e.message}`);
      return null;
    }
  }

  _buildAuthorEnv(commit) {
    const originalDate = new Date(parseInt(commit.date) * 1000).toISOString();
    return {
      GIT_AUTHOR_DATE: originalDate,
      GIT_COMMITTER_DATE: originalDate,
      GIT_AUTHOR_NAME: this.config.userName,
      GIT_AUTHOR_EMAIL: this.config.userEmail,
      GIT_COMMITTER_NAME: this.config.userName,
      GIT_COMMITTER_EMAIL: this.config.userEmail,
    };
  }

  async _applyPatchInTarget(patch, commit) {
    try {
      const patchPath = path.join(os.tmpdir(), `sync-patch-${commit.hash.slice(0, 8)}.patch`);
      fs.writeFileSync(patchPath, patch);

      try {
        await this.targetGit.raw(['apply', patchPath]);
      } catch (applyErr) {
        try {
          await this.targetGit.raw(['apply', '--3way', patchPath]);
        } catch {
          throw applyErr;
        }
      } finally {
        try { fs.unlinkSync(patchPath); } catch {}
      }

      await this.targetGit.add('-A');
      const message = commit.message.trim() || `Sync commit ${commit.hash.slice(0, 8)}`;
      const env = this._buildAuthorEnv(commit);
      await this.targetGit.env(env).commit(message);
      return true;
    } catch (e) {
      this.logger.error(`应用补丁失败 ${commit.hash.slice(0, 8)}: ${e.message}`);
      try {
        await this.targetGit.raw(['checkout', '--', '.']);
        await this.targetGit.raw(['clean', '-fd']);
      } catch {}
      return false;
    }
  }

  async _prepareTargetBranch(isFullSync) {
    if (isFullSync) {
      try {
        try {
          await this.targetGit.branch(['-D', this.config.targetBranch]);
        } catch {}

        await this.targetGit.checkout(['--orphan', this.config.targetBranch]);
        await this.targetGit.raw(['rm', '-rf', '--cached', '.']);
        await this.targetGit.clean('f', ['-d']);

        this.logger.info(`已创建孤立分支: ${this.config.targetBranch}`);
        return true;
      } catch (e) {
        this.logger.error(`创建分支失败: ${e.message}`);
        return false;
      }
    } else {
      try {
        const localBranches = await this.targetGit.branchLocal();
        if (localBranches.all.includes(this.config.targetBranch)) {
          await this.targetGit.checkout(this.config.targetBranch);
        } else {
          await this.targetGit.checkout(['-b', this.config.targetBranch, `origin/${this.config.targetBranch}`]);
        }

        this.logger.info(`已切换到目标分支: ${this.config.targetBranch}`);
        return true;
      } catch (e) {
        this.logger.error(`切换分支失败: ${e.message}`);
        return false;
      }
    }
  }

  async _recommit(commits) {
    let successCount = 0;

    for (const commit of commits) {
      const patch = await this._exportPatch(commit.hash);
      if (!patch || !patch.trim()) {
        this.logger.warning(`跳过空提交 ${commit.hash.slice(0, 8)}`);
        continue;
      }

      const ok = await this._applyPatchInTarget(patch, commit);
      if (ok) {
        successCount++;
        this.logger.info(`同步提交: ${commit.message.trim().slice(0, 50)}...`);
      }
    }

    this.logger.success(`成功同步 ${successCount}/${commits.length} 个提交`);
    return { success: true, count: successCount };
  }

  async pushToRemote() {
    try {
      const authUrl = this._getAuthUrl(this.config.targetRepoUrl);
      await this.targetGit.remote(['set-url', 'origin', authUrl]);
      await this.targetGit.push('origin', this.config.targetBranch, ['--force']);
      this.logger.success('推送到目标仓库成功');
      return true;
    } catch (e) {
      this.logger.error(`推送失败: ${e.message}`);
      return false;
    }
  }

  async sync() {
    this.logger.info('='.repeat(50));
    this.logger.info('开始跨仓库同步...');

    if (!await this.cloneOrPullSource()) return false;

    const sinceHash = this.config.lastSyncHash;
    const commits = await this.fetchCommits(sinceHash);

    if (commits.length === 0) {
      this.logger.info('没有新的提交需要同步');
      this.config.lastSyncTime = new Date().toLocaleString('zh-CN', { hour12: false });
      return true;
    }

    this.logger.info(`发现 ${commits.length} 个新提交`);

    if (!await this.cloneOrPullTarget()) return false;

    const isFullSync = !sinceHash || !await this._targetBranchExists();

    if (!await this._prepareTargetBranch(isFullSync)) return false;

    const result = await this._recommit(commits);
    if (!result.success) return false;

    if (!await this.pushToRemote()) return false;

    const lastCommit = commits[commits.length - 1];
    this.config.lastSyncHash = lastCommit.hash;
    this.config.lastSyncTime = new Date().toLocaleString('zh-CN', { hour12: false });

    this.logger.success(`跨仓库同步完成！共同步 ${result.count} 个提交`);
    this.logger.info('='.repeat(50));
    return true;
  }
}

module.exports = { GitService };
